src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -153,27 +148,25 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
             {
                 // Set some context on the CompletionItem so the CompletionResolveHandler can handle it accordingly.
                 result = SetResolveData(result.Value, serverKind);
-                if (IsSimpleImplicitExpression(documentSnapshot, request, serverKind))
+
+                var wordExtent = GetWordExtent(documentSnapshot, request, serverKind);
+                if (IsSimpleImplicitExpression(documentSnapshot, request, wordExtent, serverKind))
                 {
                     result = DoNotPreselect(result.Value);
-                    result = IncludeCSharpKeywords(result.Value, serverKind);
+                    result = IncludeCSharpKeywords(result.Value);
                 }
+
+                result = RemoveInvalidCompletionItems(result.Value, documentSnapshot, wordExtent, serverKind);
             }
 
             return result;
         }
 
-        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        private TextExtent? GetWordExtent(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
```

Comment:
I extracted the word extent logic out from `IsSimpleImplicitExpression` since the same logic also applies to removing design time items.

Reply:
This actually looks like a really good candidate for a general ITextSnapshot extension if you want to refactor it to something there.

I'd change the signature though to be more generic. Something like:

```C#
private static TextExtent? GetWordExtent(this ITextSnapshot textSnapshot, int character, int line, ITextStructureNavigatorSelectorService textStructureNavigatorService)
```